playbook/playbook_test.go

```diff
@@ -0,0 +1,201 @@
+package playbook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/flanksource/duty/fixtures/dummy"
+	"github.com/flanksource/duty/models"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	ginkgo "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = ginkgo.Describe("Playbook runner", ginkgo.Ordered, func() {
+	var (
+		playbook models.Playbook
+		runResp  RunResponse
+		consumer *queueConsumer
+	)
+
+	ginkgo.It("should store dummy data", func() {
+		dataset := dummy.GetStaticDummyData()
+		err := dataset.Populate(testDB)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	ginkgo.It("start the queue consumer in background", func() {
+		consumer = NewQueueConsumer(testDB, testDBPool)
+		go func() {
+			err := consumer.Listen()
+			Expect(err).NotTo(HaveOccurred())
+		}()
+	})
+
+	ginkgo.It("should create a new playbook", func() {
+		playbookSpec := v1.PlaybookSpec{
+			Description: "write config name to file",
+			Parameters: []v1.PlaybookParameter{
+				{Name: "path", Label: "path of the file"},
+			},
+			Approval: &v1.PlaybookApproval{
+				Type: v1.PlaybookApprovalTypeAny, // We have two approvers (John Doe & John Wick) and just a single approval is sufficient
+				Approvers: v1.PlaybookApprovers{
+					People: []string{
+						dummy.JohnDoe.ID.String(),
+						dummy.JohnWick.ID.String(),
+					},
+				},
+			},
+			Actions: []v1.PlaybookAction{
+				{
+					Name: "write config id to a file",
+					Exec: &v1.ExecAction{
+						Script: "printf {{.config.id}} > {{.params.path}}",
+					},
+				},
+				{
+					Name: "append config class to the same file ",
+					Exec: &v1.ExecAction{
+						Script: "printf {{.config.config_class}} >> {{.params.path}}",
+					},
+				},
+			},
+		}
+
+		spec, err := json.Marshal(playbookSpec)
+		Expect(err).NotTo(HaveOccurred())
+
+		playbook = models.Playbook{
+			Name:   "config name saver",
+			Spec:   spec,
+			Source: models.SourceConfigFile,
+		}
+
+		err = testDB.Create(&playbook).Error
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	ginkgo.It("should store playbook run via API", func() {
+		run := RunParams{
```

Comment:
I don't think it is necessary to call these functions via HTTP - we can all the functions directly

Reply:
Everything about saving a playbook run is handled by the HTTP handler. The actual task is simply to save to DB and the handler performs other validation on top of that.

I think its better to test the handler this way than just directly saving to db from the test.